pkg/policy/l4Filter_test.go

```diff
@@ -1153,3 +1155,80 @@ func (ds *PolicyTestSuite) TestMergingWithDifferentEndpointSelectedAllowAllL7(c
 	c.Assert(state.selectedRules, Equals, 0)
 	c.Assert(state.matchedRules, Equals, 0)
 }
+
+// Case 12: allow all at L3 in one rule with restrictions at L7. Determine that
+//          the host should always be allowed. From Host should go to proxy
+//          allow all; other L3 should restrict at L7 in a separate filter.
+func (ds *PolicyTestSuite) TestAllowingLocalhostShadowsL7(c *C) {
+
+	// This test checks that when the AllowLocalhost=always option is
+	// enabled, we always wildcard the host at L7. That means we need to
+	// set the option in the config, and of course clean up afterwards so
+	// that this test doesn't affect subsequent tests.
+	oldLocalhostOpt := option.Config.AllowLocalhost
+	option.Config.AllowLocalhost = option.AllowLocalhostAlways
+	defer func() { option.Config.AllowLocalhost = oldLocalhostOpt }()
+
+	rule := &rule{
+		Rule: api.Rule{
+			EndpointSelector: endpointSelectorA,
+			Ingress: []api.IngressRule{
+				{
+					FromEndpoints: []api.EndpointSelector{api.WildcardEndpointSelector},
+					ToPorts: []api.PortRule{{
+						Ports: []api.PortProtocol{
+							{Port: "80", Protocol: api.ProtoTCP},
+						},
+						Rules: &api.L7Rules{
+							HTTP: []api.PortRuleHTTP{
+								{Method: "GET", Path: "/"},
+							},
+						},
+					}},
+				},
+			},
+		}}
+
+	buffer := new(bytes.Buffer)
+	ctxToA := SearchContext{To: labelsA, Trace: TRACE_VERBOSE}
+	ctxToA.Logging = logging.NewLogBackend(buffer, "", 0)
+	c.Log(buffer)
+
+	expected := NewL4Policy()
+	expected.Ingress["80/TCP"] = L4Filter{
+		Port:      80,
+		Protocol:  api.ProtoTCP,
+		U8Proto:   6,
+		Endpoints: api.EndpointSelectorSlice{api.WildcardEndpointSelector},
+		L7Parser:  ParserTypeHTTP,
+		L7RulesPerEp: L7DataMap{
+			api.WildcardEndpointSelector: api.L7Rules{
```

Comment:
How is 'allow all' treated with precedence if there are more "specific" endpoint selectors? In this case, we are restricting for all endpoints, but then allowing all for the host selector. How will the datapath handle this?

Reply:
I'm not exactly sure how the precedence is determined here, but I tested this PR manually and it had the behaviour I was looking for, ie the proxy now starts seeing all traffic from the host, despite the host allowing all at L3.

/cc @rlenglet @jrajahalme Is there a precedence defined between these? Or is the wildcardendpointselector handled in a special way?

Reply:
Walked through the `getDirectionNetworkPolicy()` logic and it seems like when this L4Filter logic is applied to a set of identities there, it properly resolves the set of rules for each identity based on the list of selectors here, ie the selectors here may overlap with one another, and if multilple select a pod, then the rules are added together to determine the allowed set. If one of the selectors allows all, then it will add "allow all" to "allow something more specific", and resolve to "allow all". That's why this works.

Reply:
Awesome! :D